### PR TITLE
Add --fuzz, --real, --user to timing scripts

### LINKED ISSUE #11301: Coq's makefile.build: print-pretty-timed-diff mixes up "before" and "after"
This is #8158, but about Coq's own build system rather than coq_makefile

### PR DESCRIPTION
**Pull request by JasonGross**
- Add a `--fuzz` option to `make-both-single-timing-files.py`

  Passing `--fuzz=N` allows differences in character locations of up to `N` characters when matching lines in per-line timing diffs.

  The corresponding variable for `coq_makefile` is `TIMING_FUZZ=N`.

  See also the discussion at https://github.com/coq/coq/pull/11076#pullrequestreview-324791139

- Allow passing `--real` to per-file timing scripts and `--user` to per-line timing script.

  This allows easily comparing real times instead of user ones (or vice versa).

- Support `TIMING_SORT_BY` and `TIMING_FUZZ` in Coq's own build

- We also now use argparse rather than a hand-rolled argument parser; there were getting to be too many combinations of options.

- Fix the ordering of columns in Coq's build system; this is the equivalent of #8167 for Coq's build system.

Fixes #11301
Supersedes / closes #11022
Supersedes / closes #11230

**Kind:** documentation / bug fix / feature / infrastructure.

- [x] Added / updated test-suite
- [x] Corresponding documentation was added / updated (including any warning and error messages added / removed / modified).
- [x] Entry added in the changelog (see https://github.com/coq/coq/tree/master/doc/changelog#unreleased-changelog for details).

### CODE REVIEW

**Comment by JasonGross** [comment]
Who is willing to be assignee on this?

**Comment by ejgallego** [comment]
@jasongross thanks for the PR, I'll gladly test it, however please bear with me I think I won't be able to test it until next week.

**Comment by ejgallego** [comment]
This needs review from @coq/legacy-build-maintainers too

**Review by SkySkimmer — APPROVED** [review]
Seems OK, I didn't look carefully outside the makefile though

**Comment by JasonGross** [comment]
@ejgallego any update?

**Comment by JasonGross** [comment]
Rebased on top of master

**Comment by JasonGross** [comment]
Very strange that the test-suite fails only on Windows.  I'll have to look into this.  @MSoegtropIMC do you see anything obvious going wrong here?  (I'll report more on details as I dig into it)

**Review by Zimmi48 — COMMENTED** [review]
I don't really have time to proofread the doc, but I didn't notice anything problematic.

**Comment by JasonGross** [comment]
Windows issues should be fixed, now; I had an issue with line endings in the Python script

File: doc/sphinx/practical-tools/utilities.rst
Comment (jfehrle) [inline doc/sphinx/practical-tools/utilities.rst]:
```suggestion
       By default the table displays user times.  If the build log
```

File: doc/sphinx/practical-tools/utilities.rst
Comment (jfehrle) [inline doc/sphinx/practical-tools/utilities.rst]:
```suggestion
       Just like ``pretty-timed``, this table shows user
```

File: doc/sphinx/practical-tools/utilities.rst
Comment (jfehrle) [inline doc/sphinx/practical-tools/utilities.rst]:
```suggestion
       times by default.  Use ``TIMING_REAL=1`` on the command line to show real times instead.
```

File: doc/sphinx/practical-tools/utilities.rst
Comment (jfehrle) [inline doc/sphinx/practical-tools/utilities.rst]:
```suggestion
```

File: doc/sphinx/practical-tools/utilities.rst
Comment (jfehrle) [inline doc/sphinx/practical-tools/utilities.rst]:
```suggestion
       By default, two lines are matches only if
```

File: doc/sphinx/practical-tools/utilities.rst
Comment (jfehrle) [inline doc/sphinx/practical-tools/utilities.rst]:
```suggestion
       character offsets and initial code strings are identical.  Passing
```

File: doc/sphinx/practical-tools/utilities.rst
Comment (jfehrle) [inline doc/sphinx/practical-tools/utilities.rst]: Is this an N character prefix, suffix or any N characters scattered in the method names?  "abc" to "abcX" or "ac" or "aBc" or "acb" all match with N=1?  Order matters, eg "abc" doesn't match "cba" for N=1?

Now that I think about it just a little, this option doesn't seem very general--it probably solves your problem though probably not really cleanly.

**Review by jfehrle — COMMENTED** [review]
Minor wording suggestions and one point to clarify.

File: doc/sphinx/practical-tools/utilities.rst
Comment (JasonGross) [reply doc/sphinx/practical-tools/utilities.rst]: "defaults to using user times by default" seems worse.  What are you trying to accomplish with this suggestion?

File: doc/sphinx/practical-tools/utilities.rst
Comment (JasonGross) [reply doc/sphinx/practical-tools/utilities.rst]: This leaves the previous sentence as a fragment.  Did you mean to suggest just deleting "rather than user times"?

File: doc/sphinx/practical-tools/utilities.rst
Comment (JasonGross) [reply doc/sphinx/practical-tools/utilities.rst]: This change causes the text to miss something important, imo.  It's not just showing user times, it's also basing the computed timing differences (and the sorting) on user times.

File: doc/sphinx/practical-tools/utilities.rst
Comment (JasonGross) [reply doc/sphinx/practical-tools/utilities.rst]: Why is "are matches" better than "are considered the same"?

File: doc/sphinx/practical-tools/utilities.rst
Comment (JasonGross) [reply doc/sphinx/practical-tools/utilities.rst]: I think you missed the word "offsets".  The strings (what we can see of them) must be identical, and the line length must be identical.  Without this option, if you add one space in a comment at the beginning of the file, then every subsequent command is considered completely different due to occurring at different locations in the two versions of the document.
Perhaps the use of "characters" after "`N`" is misleading (but I don't know what else to call the units that I'm shifting the document locations by).  Do you have a suggestion of how to make this clearer?

Comment (JasonGross) [reply doc/sphinx/practical-tools/utilities.rst]: Perhaps my use of "initial code strings" is also confusing; what I meant to point at here is that only a very limited number of characters from each line appear in the log, so we don't actually have access to the full commands generating each timing.

File: doc/sphinx/practical-tools/utilities.rst
Comment (jfehrle) [reply doc/sphinx/practical-tools/utilities.rst]: Then ignore it.

File: doc/sphinx/practical-tools/utilities.rst
Comment (jfehrle) [reply doc/sphinx/practical-tools/utilities.rst]: Just for brevity/simplicity.  If you say something in fewer, simpler words without distorting the meaning, that's generally easier to understand.  However, I should have suggested "two lines match only if ...", which is even shorter and avoids the passive voice.

Change it or not as you wish.

File: doc/sphinx/practical-tools/utilities.rst
Comment (jfehrle) [reply doc/sphinx/practical-tools/utilities.rst]: Yeah, I missed "offsets".

Dropping "characters" after "N" would make it a little harder to misread.  The first part of the sentence make it clear that the units for the offsets are characters.

Why are the offsets needed?  Are the "initial code strings" not unique?  (And if there are occasionally duplicates, there's no reasonable way to qualify the names?)

The timing/performance documentation needs a couple paragraphs of overview: What files are used to collect data?  What is the "timing log"?  What data appears in the build output vs elsewhere?  It's not simple to infer that from the detail-level info.  The number of Notes suggests the material would benefit from reorganization.

Also I wonder if we should eventually do something more robust/easier to use.

Comment (JasonGross) [reply doc/sphinx/practical-tools/utilities.rst]: Initial code strings are absolutely not unique.  Consider the string `Qed.`, for example, which may show up dozens of times.

Perhaps "character locations" would be clearer than "character offsets"?  I am referring to the numbers in things like `Chars 069 - 162 [Definition~foo0~:=~Eval~comp~i...]` that occur in the output of `coqc -time`.

> What files are used to collect data?

You mean where is the data written to, or where is it taken from?

> What is the "timing log"? 

This is what I'm calling the files that I write timing information to.

> What data appears in the build output vs elsewhere?

Per-file timing data appears in the build output.  Per-line timing data is directed straight to files.

> The number of Notes suggests the material would benefit from reorganization.

Indeed.  I'm not sure how to reorganize it well, though.

> Also I wonder if we should eventually do something more robust/easier to use.

Not sure what you mean by this.  While there are a lot of fiddly options, I think the defaults are pretty good, and have been using them pretty easily for years without much trouble to get timing logs.  What gives you the sense of fragility / difficulty to use?  And what do you have in mind as a better option?

**Comment by jfehrle** [comment]
> Indeed. I'm not sure how to reorganize [the documentation] well, though.

A basic problem is that there's no introduction at all--which should say which use cases it's intended to address (e.g. a developer trying to improve performance of code internal to Coq OR a user trying to improve performance of a large proof OR).  Is the "per-line" timing only for commands or will it break out individual tactics?  And it should also answer basic questions ("What is the "timing log"?" should be answered in the doc, not in a PR).

Some of the repeated notes such "This target requires python" could be moved to the beginning and not marked as Notes.

The examples should appear before the notes for "pretty-timed", etc.  That's more important.

Group items of the same type together: print-pretty-single-time-diff and other targets should be together in the doc.  Variables should be together, too.  `TIMED` and `TIMING` options are confusingly similar.

Subheaders should be in terms of things users may want to do.  A bullet point like `TIMED=` or `print-pretty-single-time-diff` doesn't tell the reader anything unless they already know the material.

This should be related to the "Profiling Ltac tactics" section.  The different profiling methods should not be split across multiple chapters.

Using character ranges to identify lines seems very unfriendly--most editors don't have a "jump to character offset", which makes it harder for a user to find a particular line.  I also don't see why the lines are mangled with `~`.  And TIMING_FUZZ seems a very weird approach as well.

> there are a lot of fiddly options

We should try to make them non-fiddly.

It would be great if you could spend some time improving this doc.

**Comment by JasonGross** [comment]
I'm not going to have time to improve the doc in the near future.  I don't think improving the doc should block this PR, which is primarily about making the script more usable.  Some of the things you're asking for are well beyond the scope of this PR, requiring changes to coqc itself.

I agree with most of your points in principle, but I'd prefer having a separate issue tracking the documentation of this utility.  I'll open such an issue and respond to your points there.

File: doc/sphinx/practical-tools/utilities.rst
Comment (JasonGross) [reply doc/sphinx/practical-tools/utilities.rst]:
```suggestion
       character locations to differ by up to ``N``, as long
```

**Comment by jfehrle** [comment]
> I don't think improving the doc should block this PR, which is primarily about making the script more usable.

Of course not.

File: doc/sphinx/practical-tools/utilities.rst
Comment (JasonGross) [reply doc/sphinx/practical-tools/utilities.rst]:
```suggestion
       times.  Pass ``TIMING_REAL=1`` to ``make`` on the command line to show real times instead.
```

Going to go with something slightly more informative.

**Comment by JasonGross** [comment]
I now have a draft PR (#11606) on top of this, adding the ability to see memory info in the table, so I'd prefer this to get merged sooner rather than later

**Review by ejgallego — APPROVED** [review]
Thanks for your work (and patiente) @JasonGross ; indeed I agree it is time to merge this script.

A few questions come to my head w.r.t. future of the timing infrastructure; for example it would be great to have better separation of backend and user interface.

**Pull request merged**
